### ISSUE TITLE
Deepbots GoalEnv

### ISSUE THREAD

**Issue by tsampazk** [aidudezzz/deepbots#79]

**Comment by tsampazk** [comment]
@KelvinYang0320 Are you interested in taking look at gym's GoalEnv and what we need to do in order to integrate it with deepbots?

**Comment by KelvinYang0320** [comment]
@tsampazk Yes, I will take some time to look into GoalEnv.

**Comment by tsampazk** [comment]
@KelvinYang0320 Thank you!

**Comment by KelvinYang0320** [comment]
@ManosMagnus @tsampazk 
At the end of last year, [GoalEnv has been removed from Gym](https://github.com/openai/gym/pull/2516).
I think there are two alternative solutions:
1. Deepbots can still inherit GoalEnv by `pip install gym-robotics`, but users will need to install `gym-robotics` when installing `deepbots`.
2. We can copy/modify their GoalEnv and integrate it with deepbots without installing `gym-robotics`.
https://github.com/Farama-Foundation/Gym-Robotics/blob/main/gym_robotics/core.py

What do you think?

**Comment by tsampazk** [comment]
This is quite tricky i think. On the one hand, we very much prefer not to introduce any additional dependencies to the framework. Especially for a library that will be or is used by very specific use cases and not universally. On the other hand, i think that it is not proper and we shouldn't just copy the existing GoalEnv from the gym-robotics library.

@KelvinYang0320 is this actually needed for the panda example? Or any other current example?

**Comment by KelvinYang0320** [comment]
> @KelvinYang0320 is this actually needed for the panda example? Or any other current example?

I think GoalEnv is quite useful for our robotic tasks as I mentioned [here](https://github.com/aidudezzz/deepbots/pull/103#issue-1074446897).
I have made a [Panda-GoalEnv draft PR](https://github.com/aidudezzz/deepworlds/pull/52) to check these features.
1. [Goal-compatible observation space](https://github.com/KelvinYang0320/deepworlds/blob/9e98307782abbf752efa4cf1bd23ae3cb9f9487e/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/robot_supervisor_ddpg.py#L13-L43)
    * observation, desired_goal, and achieved_goal
2. [compute_reward(achieved_goal, desired_goal, info)](https://github.com/KelvinYang0320/deepworlds/blob/9e98307782abbf752efa4cf1bd23ae3cb9f9487e/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/robot_supervisor_ddpg.py#L162-L182)
    * Our `get_reward(action)` is actually not that suitable for robotics control task.

> Edit: I think i remember that some implementations of RL algorithms require GoalEnv and that's why we started working on it, right?

Yes, [HER requires the environment to inherits from `gym.GoalEnv`](https://stable-baselines3.readthedocs.io/en/master/modules/her.html#id1) in SB3.
However, it seems that they are fixing that now. 
* https://github.com/DLR-RM/stable-baselines3/pull/780

**Comment by tsampazk** [comment]
Alright thanks for the information! 
>     1. Deepbots can still inherit GoalEnv by `pip install gym-robotics`, but users will need to install `gym-robotics` when installing `deepbots`.

I thought about the additional dependency  on gym-robotics. Using this would mean adding additional inheritance to some gym-robotics class that includes GoalEnv. I will need some time to look into this further in detail, so i think we can put this on hold for a while and focus on the fixes regarding deepworlds. Unless you have a clear idea on what we can do specifically, so we can discuss it.

**Comment by KelvinYang0320** [comment]
@tsampazk  I agree to put this on hold. 👌 
Check this https://github.com/openai/gym/issues/2456 about the robotics environments.
> My goal is to make all the environments inside Gym itself good general purpose benchmarks suitable that someone new to the field of reinforcement learning can look at and say "okay, there's are the things everyone uses that I should play with and understand."

**Comment by tsampazk** [comment]
Yeah i skimmed through that issue and discussion and followed a couple of links, but it started to look like a long rabbit hole of links and discussions and just stopped myself. :P 

That quote is a really really good summary of gym's mission, but i'm not exactly sure what that means for us, being a framework for a robotic simulator. I guess we'll see in the future.

**Comment by KelvinYang0320** [comment]
@tsampazk I think we can move this issue to Pre-release 0.2.1-dev0 or close it. What do you think?

**Comment by tsampazk** [comment]
Yes @KelvinYang0320, it seems to still be an open issue, i think that we can move it to the pre-release.
